File: ui/web/src/services/group.ts
import { groupServiceOperations as ops } from '@/api/operations';
import { apiRequest } from '@/api/request';
import type { paths, components } from '@/api/schema/system-group-v1-service';
import type { RequestParameters } from '@/api/helper';

// Re-export schema types for backward compatibility
export type GroupStatus = components['schemas']['v1GroupStatus'];
export type Group = components['schemas']['v1Group'];
/** One role the group holds, with when the grant ends. */
export type RoleGrant = components['schemas']['v1RoleGrant'];
export type ListGroupsResponse = components['schemas']['v1ListGroupResponse'];
export type GetGroupResponse = components['schemas']['v1GetGroupResponse'];
export type CreateGroupRequest = components['schemas']['v1CreateGroupRequest'];
export type CreateGroupResponse = components['schemas']['v1CreateGroupResponse'];
export type UpdateGroupRequest = components['schemas']['GroupServiceUpdateGroupBody'];
export type UpdateGroupResponse = components['schemas']['v1UpdateGroupResponse'];
export type AddRolesToGroupRequest = components['schemas']['GroupServiceAddRolesToGroupBody'];
export type AddRolesToGroupResponse = components['schemas']['v1AddRolesToGroupResponse'];
export type RemoveRolesFromGroupRequest = components['schemas']['GroupServiceRemoveRolesFromGroupBody'];
export type RemoveRolesFromGroupResponse = components['schemas']['v1RemoveRolesFromGroupResponse'];
export type AddUsersToGroupRequest = components['schemas']['GroupServiceAddUsersToGroupBody'];
export type AddUsersToGroupResponse = components['schemas']['v1AddUsersToGroupResponse'];
export type RemoveUsersFromGroupRequest = components['schemas']['GroupServiceRemoveUsersFromGroupBody'];
export type RemoveUsersFromGroupResponse = components['schemas']['v1RemoveUsersFromGroupResponse'];
export type DeleteGroupResponse = components['schemas']['v1DeleteGroupResponse'];

// Helper type for list groups parameters (query)
export type ListGroupsParams = RequestParameters<paths, '/api/v1/groups', 'get'>;

export const groupService = {
  listGroups: (params?: ListGroupsParams) => apiRequest<ListGroupsResponse>(ops.listGroup, { query: params }),
  getGroup: (id: string) => apiRequest<GetGroupResponse>(ops.getGroup, { path: { id } }),
  createGroup: (data: CreateGroupRequest) => apiRequest<CreateGroupResponse>(ops.createGroup, { body: data }),
  updateGroup: (id: string, data: UpdateGroupRequest) => apiRequest<UpdateGroupResponse>(ops.updateGroup, { path: { id }, body: data }),
  deleteGroup: (id: string) => apiRequest<DeleteGroupResponse>(ops.deleteGroup, { path: { id } }),

  addRoles: (id: string, data: AddRolesToGroupRequest) => apiRequest<AddRolesToGroupResponse>(ops.addRolesToGroup, { path: { id }, body: data }),
  removeRoles: (id: string, data: RemoveRolesFromGroupRequest) => apiRequest<RemoveRolesFromGroupResponse>(ops.removeRolesFromGroup, { path: { id }, body: data }),
  addUsers: (id: string, data: AddUsersToGroupRequest) => apiRequest<AddUsersToGroupResponse>(ops.addUsersToGroup, { path: { id }, body: data }),
  removeUsers: (id: string, data: RemoveUsersFromGroupRequest) => apiRequest<RemoveUsersFromGroupResponse>(ops.removeUsersFromGroup, { path: { id }, body: data }),
};

File: server/internal/interface/grpc/handler/system/group.go
package system

import (
	"context"

	"github.com/linzhengen/hub/server/internal/usecase/system"

	"github.com/linzhengen/hub/server/internal/domain/system/group"
	pbv1 "github.com/linzhengen/hub/server/pb/system/group/v1"
)

func NewGroupHandler(
	groupUseCase system.GroupUseCase,
) pbv1.GroupServiceServer {
	return &groupHandler{groupUseCase: groupUseCase}
}

type groupHandler struct {
	groupUseCase system.GroupUseCase
}

func (h groupHandler) CreateGroup(ctx context.Context, request *pbv1.CreateGroupRequest) (*pbv1.CreateGroupResponse, error) {
	g, err := h.groupUseCase.Create(ctx, group.Factory(
		request.Name,
		request.Description,
	))
	if err != nil {
		return nil, err
	}
	return &pbv1.CreateGroupResponse{Group: groupDomainToPb(g)}, nil
}

func (h groupHandler) DeleteGroup(ctx context.Context, request *pbv1.DeleteGroupRequest) (*pbv1.DeleteGroupResponse, error) {
	if err := h.groupUseCase.Delete(ctx, request.Id); err != nil {
		return nil, err
	}
	return &pbv1.DeleteGroupResponse{}, nil
}

func (h groupHandler) GetGroup(ctx context.Context, request *pbv1.GetGroupRequest) (*pbv1.GetGroupResponse, error) {
	g, err := h.groupUseCase.Get(ctx, request.Id)
	if err != nil {
		return nil, err
	}
	return &pbv1.GetGroupResponse{Group: groupDomainToPb(g)}, nil
}

func (h groupHandler) ListGroup(ctx context.Context, request *pbv1.ListGroupRequest) (*pbv1.ListGroupResponse, error) {
	status, _ := toGroupDomainStatus(request.Status)
	items, total, err := h.groupUseCase.List(ctx, &system.ListGroupQueryParams{
		Limit:     request.Limit,
		Offset:    request.Offset,
		GroupIds:  request.GroupIds,
		GroupName: request.GroupName,
		Status:    status,
		RoleIds:   request.RoleIds,
	})
	if err != nil {
		return nil, err
	}
	var pbItems []*pbv1.Group
	for _, item := range items {
		pbItems = append(pbItems, groupDomainToPb(item))
	}
	return &pbv1.ListGroupResponse{
		Groups: pbItems,
		Total:  total,
	}, nil
}

func (h groupHandler) UpdateGroup(ctx context.Context, request *pbv1.UpdateGroupRequest) (*pbv1.UpdateGroupResponse, error) {
	status, err := toGroupDomainStatus(request.Status)
	if err != nil {
		return nil, err
	}
	g, err := h.groupUseCase.Update(ctx, &group.Group{
		Id:          request.Id,
		Name:        request.Name,
		Description: request.Description,
		Status:      status,
	})
	if err != nil {
		return nil, err
	}
	return &pbv1.UpdateGroupResponse{Group: groupDomainToPb(g)}, nil
}

func (h groupHandler) AddRolesToGroup(ctx context.Context, request *pbv1.AddRolesToGroupRequest) (*pbv1.AddRolesToGroupResponse, error) {
	g, err := h.groupUseCase.AddRoles(ctx, request.Id, request.RoleIds, timestampPbToTimePtr(request.ExpiresAt))
	if err != nil {
		return nil, err
	}
	return &pbv1.AddRolesToGroupResponse{Group: groupDomainToPb(g)}, nil
}

func (h groupHandler) RemoveRolesFromGroup(ctx context.Context, request *pbv1.RemoveRolesFromGroupRequest) (*pbv1.RemoveRolesFromGroupResponse, error) {
	g, err := h.groupUseCase.RemoveRoles(ctx, request.Id, request.RoleIds)
	if err != nil {
		return nil, err
	}
	return &pbv1.RemoveRolesFromGroupResponse{Group: groupDomainToPb(g)}, nil
}

func (h groupHandler) AddUsersToGroup(ctx context.Context, request *pbv1.AddUsersToGroupRequest) (*pbv1.AddUsersToGroupResponse, error) {
	g, err := h.groupUseCase.AddUsers(ctx, request.Id, request.UserIds, timestampPbToTimePtr(request.ExpiresAt))
	if err != nil {
		return nil, err
	}
	return &pbv1.AddUsersToGroupResponse{Group: groupDomainToPb(g)}, nil
}

func (h groupHandler) RemoveUsersFromGroup(ctx context.Context, request *pbv1.RemoveUsersFromGroupRequest) (*pbv1.RemoveUsersFromGroupResponse, error) {
	g, err := h.groupUseCase.RemoveUsers(ctx, request.Id, request.UserIds)
	if err != nil {
		return nil, err
	}
	return &pbv1.RemoveUsersFromGroupResponse{Group: groupDomainToPb(g)}, nil
}
